python-package/securexgboost/remote_server.py

```diff
@@ -229,121 +230,130 @@ def invoke(self):
                 exceptions = [result.status.exception for result in results]
                 i = statuses.index(-1)
                 exception = exceptions[i]
+
+            # Collect all signatures
+            sig_protos = []
+            sig_lens = []
+            for result in results:
+                sig_protos.append(result.signature)
+                sig_lens.append(result.sig_len)
+            
+            # If we return only one signature, return the signature from the master enclave
+            master_signature = sig_protos[0]
+            master_sig_len = sig_lens[0]
                 
             # Set return value
             if self._func == rabit_remote_api.RabitInit:
                 if error:
                     self._ret = remote_pb2.Status(status=-1, exception=exception)
                 else:
+                    # FIXME: add signatures
                     self._ret = remote_pb2.Status(status=0)
             elif self._func == rabit_remote_api.RabitFinalize:
                 if error:
                     self._ret = remote_pb2.Status(status=-1, exception=exception)
                 else:
+                    # FIXME: add signatures
                     self._ret = remote_pb2.Status(status=0)
             elif self._func == remote_api.XGDMatrixCreateFromEncryptedFile:
                 if error:
-                    self._ret = (None, remote_pb2.Status(status=-1, exception=exception)) 
+                    self._ret = (None, None, None, remote_pb2.Status(status=-1, exception=exception)) 
                 else:
                     dmatrix_handles = [result.name for result in results]
                     if dmatrix_handles.count(dmatrix_handles[0]) == len(dmatrix_handles):
                         # Every enclave returned the same handle string
-                        self._ret = (dmatrix_handles[0], remote_pb2.Status(status=0))
+                        self._ret = (dmatrix_handles[0], master_signature, master_sig_len, remote_pb2.Status(status=0))
                     else:
-                        self._ret = (None, remote_pb2.Status(status=-1, exception="ERROR: Inconsistent dmatrix handles returned by enclaves in XGDMatrixCreateFromEncryptedFile call"))
+                        self._ret = (None, None, None, remote_pb2.Status(status=-1, exception="ERROR: Inconsistent dmatrix handles returned by enclaves in XGDMatrixCreateFromEncryptedFile call"))
             elif self._func == remote_api.XGBoosterSetParam:
                 if error:
-                    self._ret = remote_pb2.Status(status=-1, exception=exception)
+                    self._ret = (None, None, remote_pb2.Status(status=-1, exception=exception))
                 else:
-                    self._ret = remote_pb2.Status(status=0)
+                    self._ret = (None, None, remote_pb2.Status(status=0))
```

Comment:
Instead of `None`, should return the master signature instead?

Reply:
Fixed